core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1053,6 +1085,22 @@ void releaseFetchLock() {
         fetchLock.set(false);
     }
 
+    /**
+     * Marks the share partition as fenced.
+     */
+    void markFenced() {
+        lock.writeLock().lock();
+        try {
+            partitionState = SharePartitionState.FENCED;
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private boolean stateNotActive() {
+        return  partitionState() != SharePartitionState.ACTIVE;
```

Comment:
We probably should throw a fenced exception and let the caller handle it. This can be done in a separate PR.

Reply:
Sure, https://issues.apache.org/jira/browse/KAFKA-17901.